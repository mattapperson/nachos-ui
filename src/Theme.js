import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
