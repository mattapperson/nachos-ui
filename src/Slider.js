import React, { Component } from "react";
import PropTypes from "prop-types";
import { View, PanResponder, Platform } from "react-native";
import Progress from "./Progress";
import { withTheme } from "./Theme";

class Slider extends Component {
  static themeConfig = {
    style: {
      base: {
        position: "relative",
        marginVertical: 25
      },
      finger: {
        backgroundColor: "rgba(0, 0, 0, 0)",
        borderRadius: 90,
        width: 72,
        height: 72
      },
      knob: {
        position: "absolute",
        top: -22 + 2,
        left: -22,
        width: 44,
        height: 44,
        borderRadius: 22,
        backgroundColor: "white",
        shadowColor: "black",
        shadowOpacity: 0.25,
        shadowRadius: 4,
        shadowOffset: { width: 0, height: 2 },
        alignItems: "center",
        justifyContent: "center",
        ...(Platform.OS === "web"
          ? {
              boxShadow: "0 2px 4px rgba(0, 0, 0, 0.25)"
            }
          : {}),
        ...(Platform.OS === "android"
          ? {
              elevation: 1
            }
          : {})
      }
    }
  };
  static propTypes = {
    value: PropTypes.number,
    maxValue: PropTypes.number,
    minValue: PropTypes.number,
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    knobStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    width: PropTypes.number,
    height: PropTypes.number,
    onValueChange: PropTypes.func,
    theme: PropTypes.object
  };

  static defaultProps = {
    value: 0,
    maxValue: 1,
    minValue: 0,
    width: 300,
    height: 6,
    onValueChange: () => {}
  };

  componentWillMount() {
    this._panResponder = PanResponder.create({
      onPanResponderTerminationRequest: () => false,
      onStartShouldSetPanResponder: () => true,
      onStartShouldSetPanResponderCapture: () => true,
      onMoveShouldSetPanResponder: () => false,
      onPanResponderGrant: this._handlePanResponderGrant,
      onPanResponderMove: this._handlePanResponderMove
    });
  }

  _handlePanResponderGrant = () => {
    this._onGrantValue = this.props.value;
  };

  _handlePanResponderMove = (e, gesture) => {
    const { onValueChange, minValue, maxValue, width } = this.props;
    const oldProgress = this._onGrantValue / (maxValue - minValue);
    const valueRange = maxValue - minValue;
    const newProgress = oldProgress + gesture.dx / width;
    const newValue = minValue + newProgress * valueRange;

    onValueChange(Math.max(minValue, Math.min(maxValue, newValue)));
  };

  render() {
    const {
      value,
      maxValue,
      minValue,
      style,
      knobStyle,
      width,
      height,
      theme
    } = this.props;

    const progress = value / (maxValue - minValue);
    const knobTransform = {
      transform: [{ translateX: width * progress }]
    };

    return (
      <View style={[theme.base, style, { width, height }]}>
        <Progress width={width} height={height} progress={progress} />
        <View
          {...this._panResponder.panHandlers}
          style={[theme.knob, knobTransform, knobStyle]}
        >
          <View style={theme.finger} />
        </View>
      </View>
    );
  }
}

export default withTheme("Slider", Slider);
