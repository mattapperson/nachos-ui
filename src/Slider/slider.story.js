import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import React from "react";
import Slider from "./index";

import { View } from "react-native";

const Example = () => {
  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Slider value={0.4} />
    </View>
  );
};

storiesOf("Slider", module).add("default", withInfo()(Example));

export default Example;
