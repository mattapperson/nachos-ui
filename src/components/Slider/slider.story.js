import { storiesOf } from "../../storybook";
import { withInfo } from "../../storybook";
import React from "react";
import Slider from "./index";

import { View } from "react-native";

const Example = () => {
  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Slider value={0.4} />
    </View>
  );
};

storiesOf("Slider", module).add("default", withInfo()(Example));

export default Example;
