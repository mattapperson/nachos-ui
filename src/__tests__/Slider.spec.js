import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import Slider from "../Slider";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Slider value={0.5} />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
