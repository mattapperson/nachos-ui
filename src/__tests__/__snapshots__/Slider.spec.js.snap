exports[`test should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "marginVertical": 25,
        "position": "relative",
      },
      undefined,
      Object {
        "height": 6,
        "width": 300,
      },
    ]
  }>
  <View
    style={
      Object {
        "backgroundColor": "#bdc1cc",
        "borderRadius": 3,
        "height": 6,
        "overflow": "hidden",
        "width": 300,
      }
    }>
    <View
      style={
        Object {
          "backgroundColor": "#2f8cff",
          "borderRadius": 3,
          "height": 6,
          "width": 150,
        }
      } />
  </View>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 22,
          "height": 44,
          "justifyContent": "center",
          "left": -22,
          "position": "absolute",
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
          "top": -20,
          "width": 44,
        },
        Object {
          "transform": Array [
            Object {
              "translateX": 150,
            },
          ],
        },
        undefined,
      ]
    }>
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "borderRadius": 90,
          "height": 72,
          "width": 72,
        }
      } />
  </View>
</View>
`;
